hazelcast/src/main/java/com/hazelcast/map/impl/event/MapEventPublisherImpl.java

```diff
@@ -140,75 +158,114 @@ public void publishEvent(Address caller, String mapName, EntryEventType eventTyp
             return;
         }
 
-        List<EventRegistration> includeValueRegistrations = null;
-        List<EventRegistration> nullValueRegistrations = null;
+        publishEvent(registrations, caller, mapName, eventType, dataKey, oldValue, value, mergingValue);
+    }
+
+    private void publishEvent(Collection<EventRegistration> registrations, Address caller, String mapName,
+                                     EntryEventType eventType, Data dataKey, Object oldValue, Object value,
+                                     Object mergingValue) {
+
+        Map<Integer, EntryEventData> eventDataIncludingValues = null;
+        Map<Integer, EntryEventData> eventDataExcludingValues = null;
```

Comment:
isn't an `ArrayList` cheaper than `Map` to hold event data in terms of number of created objects and calculations needed for hash? Seems an `ArrayList` is a better fit here and no need a Map